[Unified TorchTrainer] Add HF Transformers TorchTrainer Utilities (#38083)
ray-project/ray: doc/source/train/user-guides/checkpoints.rst

        latest saved checkpoint.

        .. code-block:: python
            :emphasize-lines: 11

            from ray.train.huggingface import (
                RayTrainReportCallback,
                prepare_trainer
            )
            from ray.train.torch import TorchTrainer

            def train_func(config):
                ...

                trainer = transformers.Trainer(...)

                trainer.add_callback(RayTrainReportCallback())
                
                trainer = prepare_trainer(trainer)
                trainer.train()
            
            ray_trainer = TorchTrainer(
                train_func,
                run_config=RunConfig(
                    checkpoint_config=CheckpointConfig(
                        num_to_keep=3,
                        checkpoint_score_attribute="eval_loss",
                        checkpoint_score_order="min",
                    )
                )
            )
        
        You should properly configure the `logging_strategy`, `save_strategy` 
        and `evaluation_strategy`, so that at the checkpoint saving step, transformers 

>         trainer also reports the latest monitoring metrics (e.g. `eval_loss` in the above case).

Reviewer: I think it would be helpful to include these directly in the example itself, as it's currently not being shown. 

In other words, the example should make it clear that our callback is complementary to Transformers logging/checkpointing.

Author: Make sense. Like explicitly define the `save_strategy` and `evaluation_strategy`, and put these info as comments in the example?

        For example, if we want to keep the top-k checkpoint regards to metric `eval_loss`, 
        you need to make sure that evaluation happens at the same step of checkpoint saving.

        Several valid strategy settings are:
        - `save_strategy = evaluation_strategy = "epoch"`
        - `save_strategy = evaluation_strategy = "steps"`, `save_steps % eval_steps == 0`

        Several invalid configurations can be:
        - Set different strategies: `save_strategy != evaluation_strategy`
        - Set the same strategy but with mismatched frequency: `save_steps % eval_steps != 0` 

        **Option 2: Implement your customized report callback**

>         If you feel that Ray Train's default report callbacks are not sufficient for your use case, 
>         implement a callback yourselve! Below is a example implementation that collects latest metrics 
>         and reports on checkpoint save.

Reviewer: ```suggestion
        If you feel that Ray Train's default `RayTrainReportCallback` is not sufficient for your use case, you can also
        implement a callback yourself! Below is a example implementation that collects latest metrics 
        and reports on checkpoint save.
```

                    """Log is called on evaluation step and logging step."""
                    self.metrics.update(logs)

                def on_save(self, args, state, control, **kwargs):
                    """Event called after a checkpoint save."""
                    with TemporaryDirectory() as tmpdir:
                        # Copy the latest checkpoint to tempdir
                        source_ckpt_path = transformers.trainer.get_last_checkpoint(args.output_dir)
                        target_ckpt_path = os.path.join(tmpdir, "checkpoint")
                        shutil.copytree(source_ckpt_path, target_ckpt_path)

                        # Build Ray Train Checkpoint
                        checkpoint = Checkpoint.from_directory(tmpdir)

                        # Report to Ray Train with up-to-date metrics
                        ray.train.report(metrics=self.metrics, checkpoint=checkpoint)

                        # Clear the metrics buffer
                        self.metrics = {}

        You can customize when(`on_save`, `on_epoch_end`, `on_evaluate`) and 
        what(customized metrics and checkpoint files) to report by implementing your own 

>         transformers trainer callback.

Reviewer: ```suggestion
        Transformers Trainer callback.
```

                # `evaluation_strategy` to ensure that the monitoring metrics
                # are always logged at the same step of checkpoint saving.

                # Valid configurations for eval metrics
                args = TrainingArguments(
                    ...,
                    evaluation_strategy="epoch",
                    save_strategy="epoch",
                )

                args = TrainingArguments(
                    ...,
                    evaluation_strategy="steps",
                    save_strategy="steps",
                    eval_steps=100,
                    save_steps=100,
                )

                args = TrainingArguments(
                    ...,
                    evaluation_strategy="steps",
                    save_strategy="steps",
                    eval_steps=50,
                    save_steps=100,
                )

                # Invalid configuration for eval metrics
                invalid_args = TrainingArguments(
                    ...,
                    evaluation_strategy="steps",
                    save_strategy="steps",
                    save_steps=100,
                    eval_steps=70,

>                 )

Reviewer: Oh I would exclude these from this code block. In this code snippet we can leave the first one in here, and have a comment above with something like:

```
# Configure logging_strategy  and saving_strategy as you normally would
```

The idea is that the focus of this block is to show the user that our callback is complementary with Transformers, and then a second section could  be for tips on ensuring compatibility.

Author: oh i see. That makes sense. We don't want to restrict the usability of the original transformers trainer.

Updated!
